target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,750 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Realtek RTL83XX clock driver
+ * Copyright (C) 2022 Markus Stockhausen
+ *
+ * This driver provides basic clock support for the central core clock unit (CCU) and its PLLs
+ * inside the RTL838X and RTL8389X SOC. Currently CPU, memory and LXB clock information can be
+ * accessed. To make use of the driver add the following devices and configurations at the
+ * appropriate locations to the DT.
+ *
+ * #include <dt-bindings/clock/rtl83xx-clk.h>
+ *
+ * sram0: sram@9f000000 {
+ *   compatible = "mmio-sram";
+ *   reg = <0x9f000000 0x18000>;
+ *   #address-cells = <1>;
+ *   #size-cells = <1>;
+ *   ranges = <0 0x9f000000 0x18000>;
+ * };
+ *
+ * osc: oscillator {
+ *   compatible = "fixed-clock";
+ *   #clock-cells = <0>;
+ *   clock-frequency = <25000000>;
+ * };
+ *
+ * ccu: clock-controller {
+ *   compatible = "realtek,rtl8380-clock";
+ *   #clock-cells = <1>;
+ *   clocks = <&osc>;
+ *   clock-names = "ref_clk";
+ *   cpu-range = <MIN MAX>;
+ *   mem-range = <MIN MAX>;
+ *   lxb-range = <MIN MAX>;
```

Comment:
Unless I missed something, these are non-standard properties. Why can't these limits just go in the driver? Do you want to be able to limit the clock frequencies per board?

Reply:
Absolutely right. Simple example for possible differences:

Zyxel GS1900-48: Board: RTL839x CPU:750MHz LXB:200MHz MEM:400MHz
DGS 1210-52: Board: RTL839x CPU:700MHz LXB:200MHz MEM:400MHz

Reply:
That made me realise the GS1900-48 is already running faster than the 700MHz listed on the RTL8393's product page. Still, these frequencies are set up by the bootloader. For some reason, be it info from Realtek or their own testing, these boosted clock frequencies are considered stable. Would it make sense to use these boot values as an upper limit in the driver? The lower limit could still be hardcoded, as I don't think those would be determined by the boards.

Reply:
I will go with the following naming convention in the final version. max values will be possible but not set.

```
        ccu: clock-controller {
                compatible = "realtek,rtl8390-clock";
                #clock-cells = <1>;
                clocks = <&osc>;
                clock-names = "ref_clk";
                min-cpu-frequency = <425000000>;
                min-mem-frequency = <325000000>;
                min-lxb-frequency = <100000000>;
        };
```

Reply:
I'm still not convinced these min/max values belong in the DT description. Debugfs entries to configure the allowed range would make sense to me, which would be initialised by the driver to safe values (hard coded or derived from boot state).

Reply:
"normally" the hardware has internal defaults. e.g. the chip refuses to operate outside of X specs, because those are design limitations. Those do not belong int he dt, but in the driver itself.

Now, we have certain boards that can (but should they? maybe leave it as stock and loose the OC ability ...) be over/underclocked outside of spec.

For the MVP/initial series, leave that complexity out. Better have all boards behave the same, without the loss of the boost, for simpler and easier code.

We can then later argue about what the best approach is to boost' frequencies on boards that support it. But this is 'basic overclocking' like stuff. Is it worth it? Probably not. Should we facilitate it, maybe. How? That's the tricky one. If you look at common desktop CPU's, this is handled in completely different ways. But lets keep that out of the discussion for now? For upper bounds, you can always make it so, unless explicitly set higher by the user, use the internal 'max' as max?

Reply:
already fixed